3party/opening_hours/opening_hours.cpp

```diff
@@ -0,0 +1,1384 @@
+      ost << ':';
+      PrintPaddedNumber(ost, std::abs(minutes), 2);
+      ost << ')';
+    }
+    else
+      ost << time.GetEvent();
+  }
+  else if (time.IsMinutes())
+    PrintPaddedNumber(ost, std::abs(minutes), 2);
+  else
+  {
+    PrintPaddedNumber(ost, std::abs(hours), 2);
+    ost << ':';
+    PrintPaddedNumber(ost, std::abs(minutes), 2);
+  }
+
+  return ost;
+}
+
+
+bool Timespan::IsEmpty() const
+{
+  return !HasStart() && !HasEnd();
+}
+
+bool Timespan::IsOpen() const
+{
+  return HasStart() && !HasEnd();
+}
+
+bool Timespan::HasStart() const
+{
+  return GetStart().HasValue();
+}
+
+bool Timespan::HasEnd() const
+{
+  return GetEnd().HasValue();
+}
+
+bool Timespan::HasPlus() const
+{
+  return m_plus;
+}
+
+bool Timespan::HasPeriod() const
+{
+  return m_period.HasValue();
+}
+
+Time const & Timespan::GetStart() const
+{
+  return m_start;
+}
+
+Time const & Timespan::GetEnd() const
+{
+  return m_end;
+}
+
+Time const & Timespan::GetPeriod() const
+{
+  return m_period;
+}
+
+void Timespan::SetStart(Time const & start)
+{
+  m_start = start;
+}
+
+void Timespan::SetEnd(Time const & end)
+{
+  m_end = end;
+}
+
+void Timespan::SetPeriod(Time const & period)
+{
+  m_period = period;
+}
+
+void Timespan::SetPlus(bool const plus)
+{
+  m_plus = plus;
+}
+
+bool Timespan::IsValid() const
+{
+  return false; // TODO(mgsergio): implement validator
+}
+
+std::ostream & operator<<(std::ostream & ost, Timespan const & span)
+{
+  ost << span.GetStart();
+  if (!span.IsOpen())
+  {
+    ost << '-' << span.GetEnd();
+    if (span.HasPeriod())
+      ost << '/' << span.GetPeriod();
+  }
+  if (span.HasPlus())
+    ost << '+';
+  return ost;
+}
+
+std::ostream & operator<<(std::ostream & ost, osmoh::TTimespans const & timespans)
+{
+  PrintVector(ost, timespans);
+  return ost;
+}
+
+
+bool NthWeekdayOfTheMonthEntry::IsEmpty() const
+{
+  return !HasStart() && !HasEnd();
+}
+
+bool NthWeekdayOfTheMonthEntry::HasStart() const
+{
+  return GetStart() != NthDayOfTheMonth::None;
+}
+
+bool NthWeekdayOfTheMonthEntry::HasEnd() const
+{
+  return GetEnd() != NthDayOfTheMonth::None;
+}
+
+NthWeekdayOfTheMonthEntry::NthDayOfTheMonth NthWeekdayOfTheMonthEntry::GetStart() const
+{
+  return m_start;
+}
+
+NthWeekdayOfTheMonthEntry::NthDayOfTheMonth NthWeekdayOfTheMonthEntry::GetEnd() const
+{
+  return m_end;
+}
+
+void NthWeekdayOfTheMonthEntry::SetStart(NthDayOfTheMonth const s)
+{
+  m_start = s;
+}
+
+void NthWeekdayOfTheMonthEntry::SetEnd(NthDayOfTheMonth const e)
+{
+  m_end = e;
+}
+
+std::ostream & operator<<(std::ostream & ost, NthWeekdayOfTheMonthEntry const entry)
+{
+  if (entry.HasStart())
+    ost << static_cast<uint32_t>(entry.GetStart());
+  if (entry.HasEnd())
+    ost << '-' << static_cast<uint32_t>(entry.GetEnd());
+  return ost;
+}
+
+bool WeekdayRange::HasWday(Weekday const & wday) const
+{
+  if (IsEmpty() || wday == Weekday::None)
+    return false;
+
+  if (!HasEnd())
+    return GetStart() == wday;
```

Comment:
It's not-obvious that WeekdayRange without End() represents a single day. Maybe it's worth for single days set End() to the same value as Start()?